apis/kahu/v1beta1/backup.go

```diff
@@ -0,0 +1,174 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+type Backup struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   BackupSpec   `json:"spec,omitempty"`
+	Status BackupStatus `json:"status,omitempty"`
+}
+
+type BackupSpec struct {
+	// +optional
+	ReclaimPolicy ReclaimPolicyType `json:"reclaimPolicy,omitempty"`
+
+	// +nullable
+	// +optional
+	MetadataLocation *BackupLocation `json:"metadataLocation,omitempty"`
+
+	// +optional
+	PreExecHook ResourceHookSpec `json:"preExecHook,omitempty"`
+
+	// +optional
+	PostExecHook ResourceHookSpec `json:"postExecHook,omitempty"`
+
+	// +optional
+	IncludedProviders []string `json:"includedProviders,omitempty"`
+
+	// +optional
+	ExcludeProviders []string `json:"excludedProviders,omitempty"`
+
+	// +optional
+	EnableMetadataBackup bool `json:"enableMetadataBackup,omitempty"`
+
+	// +optional
+	EnableVolumeBackup bool `json:"enableVolumeBackup,omitempty"`
+
+	// +optional
+	IncludedNamespaces []string `json:"includedNamespaces,omitempty"`
+
+	// +optional
+	ExcludedNamespaces []string `json:"excludedNamespaces,omitempty"`
+
+	// +optional
+	IncludedResources []string `json:"includedResources,omitempty"`
+
+	// +optional
+	ExcludedResources []string `json:"excludedResources,omitempty"`
+
+	// +optional
+	Label *metav1.LabelSelector `json:"label,omitempty"`
+}
+
+type ResourceHookSpec struct {
+	// +optional
+	Name string `json:"name"`
+
+	// +optional
+	IncludedNamespaces []string `json:"includedNamespaces,omitempty"`
+
+	// +optional
+	ExcludedNamespaces []string `json:"excludedNamespaces,omitempty"`
+
+	// +optional
+	IncludedResources []string `json:"includedResources,omitempty"`
+
+	// +optional
+	ExcludedResources []string `json:"excludedResources,omitempty"`
+
+	// +optional
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+}
+
+type ReclaimPolicyType struct {
+	// +optional
+	ReclaimPolicyDelete string `json:"reclaimPolicyDelete,omitempty"`
+
+	// +optional
+	ReclaimPolicyRetain string `json:"reclaimPolicyRetain,omitempty"`
+}
+
+type BackupCondition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	ResourceName string `json:"resourceName,omitempty"`
+
+	// +optional
+	Status string `json:"status,omitempty"`
+}
+
+type BackupPhase string
+
+const (
+	// BackupPhaseNew means the backup has been created but not
+	// yet processed by the BackupController.
```

Comment:
These states to be "Init", "VolumeBackupInProgress", "VolumeBackupFailed","VolumeBackupFinished", "ResourceBackupInProgress", "ResourceBackupFailed","ResourceBackupFinished","BackupCompleted", "BackupDeleting"